### ISSUE TITLE
Use of method cache can lead to infinite recursion in providers

### ISSUE THREAD

**Issue by jejb** [openssl/openssl#20600]
If a provider implementation relies on a library that itself links to openssl, you can get into an infinite recursion scenario where the library linked openssl calls back into the provider which then calls an operation in the library which calls back into the provider and so on until the stack overflows.  I noticed this when converting the TPM2 engine to be a provider: the Trusted Security Stacks it uses (both the IBM and the Intel ones) need to perform an ECDH operation to compute an encrypted seed for session based transport security.  The problem arises if the TPM2 provider operates on an EC TPM key, thus priming the cache so the EC derivation for security calls back into the TPM2 provider to generate the EC key for the salt, but key generation also needs to perform an ECDH operation for transport security and so on.

The problem is specifically that ossl_method_construct() goes in provider order (so would find the default provider first), but this is pre-empted if there is any provider in the method cache because the cache is searched first before trying method construction occurs.  Selecting the TPM2 EC key routines for decoding and key management by operating on the key file primes the cache and thereafter every EC key operation goes through the TPM2 provider.  The only way I've found to fix this is to add this code in the decoder

		EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
		EVP_PKEY_CTX_free(ctx);

Which primes the cache with the default provider EC methods and thanks to the fact that the TPM2 provider is loaded later and has a higher address, the default methods are then retrieved from the cache first (since the cache comparison function goes in address order of provider ... and thus would fail in the face of something like address space layout randomization).  However, it would be really nice to have a way of preventing this without resorting to the above hack, say by allowing the provider to exclude itself from the cache for lower library calls.

**Comment by beldmit** [comment]
@simo5 JFYI

**Comment by simo5** [comment]
Yeah experienced this already with softhsm2 and other similar stuff.
I think libraries should always use their own library context where you could set specific providers you know you are ok using, instead of the general context.
But it would definitely be good if there was recursion protection in OpenSSL as well.

**Comment by jejb** [comment]
On Sun, 2023-03-26 at 09:13 -0700, Simo Sorce wrote:
> Yeah experienced this already with softhsm2 and other similar stuff.

Could you share your work around?  I'm really not convinced mine is
effective in all circumstances.

**Comment by paulidale** [comment]
1. A library could create its own library context and use this for selecting dependent algorithms.
2. A library could install a property query that excludes itself for the lookup "provider != 'me'.
3. You might be able to use child providers (`OSSL_LIB_CTX_new_child`) to handle this but I suspect some care would still be required.

**Comment by jejb** [comment]
On Sun, 2023-03-26 at 15:22 -0700, Pauli wrote:
> 1. A library could create its own library context and use this for
> selecting dependent algorithms.

It's not my library: it's a library my provider has to link with, so I
have no control of this

> 2. A library could install a property query that excludes itself for
> the lookup "provider != 'me'.

The problem is happening in the cache, not in the method construct. 
The cache query that causes this is "", so no override of the query
type is possible there.  Flushing the cache and forcing queries through
method construct would work, but there's no way to do that currently
(and if I could, that would be a possible solution since the next
lookup in the library would find the default provider via method
construction).

> 3. You might be able to use child providers
> (`OSSL_LIB_CTX_new_child`) to handle this but I suspect some care
> would still be required.

I did think of this, but I couldn't find a viable mechanism

James

**Comment by paulidale** [comment]
Try `EVP_set_default_properties()` around the errant fetch.

**Comment by jejb** [comment]
On Sun, 2023-03-26 at 16:16 -0700, Pauli wrote:
> Try `EVP_set_default_properties()` around the errant fetch.

My provider isn't a party to either fetch.  The first one occurs inside
the openssl app that requests the key that eventually calls my provider
and the second one occurs inside the service library my provider is
using.  To work around this openssl bug, I need to do something with
what I can control inside the provider code, which is why I hit on the
cache priming as I explained.

James

**Comment by jejb** [comment]
On Mon, 2023-03-27 at 02:16 -0700, Matt Caswell wrote:
> 
> So, is the service library just using the default lib ctx rather than
> setting up its own one?

Both the openssl app I'm using (pkeyutil) and the lower library use the
default context.

>  The ideal case would be that the service library uses its own libctx
> which doesn't then interfere with other parts of the application.

That's really not easy: I currently pick one of two libraries depending
on compile configuration, both of which grew up on openssl-1 and each
of which would need considerable surgery to use a non-default library
context. (And neither of which is my code).

> Failing that you might be able to use `OSSL_LIB_CTX_set0_default()`
> to replace the default libctx being used within the service library.
> By changing the default properties in the replacement default libctx
> to exclude your own provider then you should be able to control the
> fetches it is making:
> 
> https://www.openssl.org/docs/manmaster/man3/OSSL_LIB_CTX_set0_default.html

I think this might work if I can find the right interception point. 
The problem is the current hack merely populates the cache with the
default provider EC key handling and that interception point occurs
before pkeyutl populates it with the provider EC key handing.  Since
both the library and pkeyutl use the default cache, when I do the
library context replacement, I replace for both so it has to occur
after pkeyutl has done the cache population but before the next call to
the lower library.  In many ways, since pkeyutl will also unexpectedly
have its default library context switched on it (doesn't matter to
pkeyutl, but might matter to a more complex application),  it looks
like a bigger hack than the one I'm already using.

**Comment by mattcaswell** [comment]
> I think this might work if I can find the right interception point.

Isn't it just a case of ensuring you call `OSSL_LIB_CTX_set0_default()` immediately before you make any call to the service library (and set it back again afterwards). Of course if you are making a lot of calls then that could be a lot of places...

Or alternatively you could call it at every entry point to your provider, and set it back before your return control back to the caller.

Doing it this way effectively means that the service library is using a different libctx to the rest of the app and therefore there is no possibility for the cache population being a problem (the caches are libctx specific).

**Comment by mattcaswell** [comment]
> Probably it's worth discussing on an OTC meeting. The idea of using OpenSSL implementation for all that is not supported by a particular HW backend sounds quite natural to me.

In my mind the problem here is the use of the default libctx by multiple different components of the same application. This sort of thing was an issue in OpenSSL 1.x too if different components all used OpenSSL in competing ways. It was, at least in part, some of the motivation for introducing libctx. The long term "fix" for this is for libraries to use a dedicated libctx. But clearly that doesn't work for existing codebases built for 1.x that are not libctx aware.

**Comment by t8m** [comment]
> But clearly that doesn't work for existing codebases built for 1.x that are not libctx aware.

Yeah, on the other hand this kind of issues mostly happens when writing a provider which is a new thing in 3.0 so it is not a regression at least.

BTW engines had the same problem in fact, except if you used low-level functions from your engine, you did not encounter this.

**Comment by mattcaswell** [comment]
> @t8m when I used a custom RSA implementation in an engine, I was able to copy the default RSA method, replace several callbacks and set the resulting method as a default one.

I think you are describing a different problem to the one covered by this issue. This issue, IMO, is specific to re-entering the default libctx from provider code potentially leading to a recursion. The answer to this, IMO, is to avoid using the default libctx from within a provider if it might lead to such a recursion. The problem here is figuring out the best way to achieve that.

**Comment by jejb** [comment]
On Mon, 2023-03-27 at 06:16 -0700, Matt Caswell wrote:
> > I think this might work if I can find the right interception point.
> 
> Isn't it just a case of ensuring you call
> `OSSL_LIB_CTX_set0_default()` immediately before you make any call to
> the service library (and set it back again afterwards). Of course if
> you are making a lot of calls then that could be a lot of places...

No, because the upper application (which is an openssl one, by the way)
is also using the default context and the key problem is that it (the
upper application) is populating the cache.  If I change the default
libctx, I change it for *both* the upper application and the lower
library.  So I have to change it after the upper application has
populated the cache, but before the lower library tries to use it to do
its ECDH manipulations.  That means I have to have considerable
knowledge of both the upper application and the lower library.

> Or alternatively you could call it at every entry point to your
> provider, and set it back before your return control back to the
> caller.
> 
> Doing it this way effectively means that the service library is using
> a different libctx to the rest of the app and therefore there is no
> possibility for the cache population being a problem (the caches are
> libctx specific).

That would probably work, since it now separates the lower library from
the upper application, but having to wrapper *every* lower library call
is a huge effort and seems rather suboptimal.

**Comment by mattcaswell** [comment]
>  If I change the default
libctx, I change it for *both* the upper application and the lower
library

Err...no? What I am suggesting here is that you wrap every call to the service library by changing the deflt libctx to something else immediately before you call it, and reset it afterwards. This means that the service library is using a different default libctx to the upper application and therefore they do not interfere with each other and have different caches.

**Comment by simo5** [comment]
> On Sun, 2023-03-26 at 09:13 -0700, Simo Sorce wrote: Yeah experienced this already with softhsm2 and other similar stuff.
> Could you share your work around? I'm really not convinced mine is effective in all circumstances.

There is no workaround really, in the case of OpenSC I contributed a [patches](https://github.com/OpenSC/OpenSC/pull/2712) to set up a libctx through the library and make it use only the default (and legacy) provider by explicitly loading them. This prevents OpenSC from seeing any other provider and avoids it conflicting with the main application it is serving.

**Comment by t8m** [comment]
**OTC: we discussed this and we agree that we should make the providers authors life easier at least by better documentation and perhaps some helper functions.**
